### ISSUE TITLE
POSIX::errno unreliable

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3258]
Migrated from [rt.perl.org#5444](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5444) (status was 'resolved')

Searchable as RT5444$

**Comment by p5pRT** [comment]
### From tds@livecell.cellnet.aldigital.co.uk


\#\!/usr/bin/perl
use strict;
use POSIX qw\(&#8203;:signal\_h :errno\_h :sys\_wait\_h\);
use Test;

\# We ran into this problem using waitpid\(\)\, but it appears to be
\# generally applicable to POSIX&#8203;::errno\. The "open" system call fails\,
\# \(because "idonotexist" doesn't exist\)\, and this is consistent with 
\# $\!\, if you look at it\. However\, POSIX&#8203;::errno\, which should match
\# the numeric equivalent of $\!\, doesn't\.

\# POSIX&#8203;::errno appears to work sometimes \(if you've done other things\, such
\# as wait\(\)s\.\.\.?\) but not in an obviously predictable way\.

\# this test script seems to work under Perl5\.005\, but fails under 5\.6\.

BEGIN \{ plan tests => 1\};

my $not = "idonotexist";
\(\!\-e $not\) or die "that will confuse this test case badly";

open FOO\, $not;
my $e = POSIX&#8203;::errno\(\);
ok\($e\);

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=high

Site configuration information for perl v5.6.0:

Configured by root at Thu Nov  2 14:48:34 GMT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.6, archname=sun4-solaris
    uname='sunos livecell 5.6 generic_105181-17 sun4u sparc sunw,ultra-enterprise '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O', gccversion=
    cppflags='-I/usr/local/include'
    ccflags ='-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib '
    libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-KPIC', lddlflags='-G -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/people/tds/work/catbert-cellnet-genie-lite/lib
    /usr/people/tds/work/catbert-cellnet-genie-modules/xslib
    /usr/local/lib/perl5/5.6.0/sun4-solaris
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl/5.005/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.005
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/usr/people/tds
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/aldl/informix/lib:/usr/aldl/informix/lib/esql:/usr/dt/lib:/usr/openwin/lib:/usr/ucblib:/usr/local/lib:/usr/local/pgsql/lib:/usr/aldl/informix/lib:/usr/aldl/informix/lib/esql
    LOGDIR (unset)
    PATH=/usr/bin:/usr/sbin:/sbin:/opt/WLtop/bin:/opt/SUNWsymon/bin:/opt/SUNWsymon/sbin:/opt/SUNWvts/bin:/usr/ccs/bin:/usr/lib/lp/bin:/usr/platform/sun4u/sbin:/opt/SUNWspro/bin:/usr/local/bin:/usr/local/netscape:/opt/FSFpatch/bin:/opt/FSFtexi/bin:/opt/FSFemacs/bin:/opt/FSFbash/bin:/usr/local/ssl/bin:/usr/openwin/bin:/opt/SUNWspro/SW3.1/bin:/opt/SUNWspro/bin:/bin:/usr/bin:/usr/ucb:/usr/bin/X11:/usr/local/bin:/usr/local/bin:/usr/sbin:/usr/local/bin:/usr/X/bin:/opt/FSFgzip/bin:/usr/ccs/bin:/usr/opt/SUNWmd/sbin:/var/qmail/bin:/opt/SUNWvxva/bin:/usr/aldl/informix/bin:/usr/afs/bin:/usr/aldl/support:/usr/aldl/informix/bin:/usr/people/tds/work/catbert-cellnet-genie-lite/util:/usr/people/tds/work/catbert-cellnet-genie-modules/handoff/version_control:.
    PERL5LIB=/usr/people/tds/work/catbert-cellnet-genie-lite/lib:/usr/people/tds/work/catbert-cellnet-genie-modules/xslib
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I submitted this bug report back in January\, though it looks like my
email address got fouled up\, so apologies if anyone's tried to mail me &
got bounces or no reponse\. As far as I can gather from bugs\.perl\.org and
this list's archives\, there's not been anything on it\.

Fortunately\, I've now found out more\.

It looks like\, in some\[1\] situations\, the first call to POSIX&#8203;::errno\(\)
doesn't return the proper value\. This seems to be because POSIX uses
AutoLoader to load & compile the subroutine\, and somewhere along the
way\, $\! gets set\. POSIX&#8203;::errno just returns \($\! \+ 0\)\. This is a classic
example of What Goes Wrong with Global Variables\, and takes me right
back to my BBC Basic days\.

Obvious workaround&#8203;: just call POSIX&#8203;::errno\(\)\, which will force POSIX to
go load it\. After that\, it should work\.

As for fixes\, presumably the options are&#8203;:

1\. Patch AutoLoader\.pm\, so it puts $\! back as it found it\[2\]\.
2\. Patch POSIX\.pm\, so it doesn't expect AutoLoader to necessarily leave
$\! alone\.

Any opinions? Whichever it is\, it's probably a trivial fix \- should I
send a patch to someone?

Cheers

\-\-
Tim Sweetman
A L Digital
"Computers are ridiculous\. So is science in general\."
 \-\-\- Church\-Turing Thesis\, Theodore Roszak Version

\[1\] not sure what exactly \- all our recent \(within last 6 months?\) Perl
installations exhibit the problem\, even one which has been reverted from
5\.6 to 5\.005\, and that's BSD\, Linux and Solaris\. 

\[2\] But while we're at it\, are there other mystery globals that
AutoLoader should avoid fu^H^H messing with?

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

bash\-2\.05$ diff \-u lib/AutoLoader\.pm\- lib/AutoLoader\.pm
<details><summary>Inline Patch</summary>

```diff
--- lib/AutoLoader.pm-  Tue Jun 12 17:24:43 2001
+++ lib/AutoLoader.pm   Tue Jun 12 17:24:53 2001
@@ -84,6 +84,7 @@
        }
     }
     my $save = $@;
+    local $!;
     eval { local $SIG{__DIE__}; require $filename };
     if ($@) {
        if (substr($sub,-9) eq '::DESTROY') {


```
</details>
_Two_ test scripts:

\-\-\- first script \-\-\-
\#\!/usr/bin/perl \-w
use strict;
use POSIX qw\(errno\);
use Test;

\# POSIX&#8203;::errno is autoloaded\.
\# autoloading requires many system calls\.
\# errno\(\) looks at $\! to generate its answer\.
\# first time you call errno\(\)\, autoloading process may alter
\# the data you were looking for\!

BEGIN \{ plan tests => 1\};

my $test = 0;
$\! = $test;
print "\# we assigned to $test to \\$\!&#8203;: $\!\\n";
my $foo = $\!;
my $errno = POSIX&#8203;::errno\(\);

\# should be numerically identical\, unless $\! hit by autoload
ok\($errno == $foo\);

\-\-\- second script \-\-\-
\#\!/usr/bin/perl \-w
use strict;
use POSIX qw\(errno\);
use Test;

\# POSIX&#8203;::errno is autoloaded\.
\# autoloading requires many system calls\.
\# errno\(\) looks at $\! to generate its answer\.
\# first time you call errno\(\)\, autoloading process may alter
\# the data you were looking for\!

BEGIN \{ plan tests => 1\};

my $test = 1;
$\! = $test;
print "\# we assigned to $test to \\$\!&#8203;: $\!\\n";
my $foo = $\!;
my $errno = POSIX&#8203;::errno\(\);

\# should be numerically identical\, unless $\! hit by autoload
ok\($errno == $foo\);
\-\-\- end \-\-\-
